### PR TITLE
UI: Add preview scaling options through edit and context menus. Allow scrolling of the preview when using a fixed scale by holding spacebar + click-n-drag.

### PR DESCRIPTION
**Pull request by Savestate2A03**
A feature I missed from the original OBS was a toggle option for 1:1 scaling for the preview. I believe I hit all the marks for this, but if I messed up with anything, I'll be happy to fix it up.

### CODE REVIEW

**Comment by RytoEX** [comment]
You should probably squash your commits into one commit.

**Comment by Savestate2A03** [comment]
Wasn't aware I could do that! Still moderately new to GitHub.

**Comment by RytoEX** [comment]
I forgot to suggest - you should probably adjust your commit log message to indicate what component was changed.  Something like "UI: Allow enable/disable of preview scaling" would probably work.

**Comment by Savestate2A03** [comment]
Ah, can do!

File: UI/window-basic-main.cpp [hunk 1]
```diff
@@ -2573,10 +2578,17 @@ void OBSBasic::ResizePreview(uint32_t cx, uint32_t cy)
 
 	/* resize preview panel to fix to the top section of the window */
 	targetSize = GetPixelSize(ui->preview);
-	GetScaleAndCenterPos(int(cx), int(cy),
-			targetSize.width()  - PREVIEW_EDGE_SIZE * 2,
+
+	if (ui->preview->Scaled()) {
+		GetScaleAndCenterPos(int(cx), int(cy),
+			targetSize.width() - PREVIEW_EDGE_SIZE * 2,
```
Comment (RytoEX) [inline UI/window-basic-main.cpp]: Tab in twice for statement continuations (lines 2584-2586).

File: UI/window-basic-main.cpp [hunk 2]
```diff
@@ -3089,6 +3101,12 @@ void OBSBasic::CreateSourcePopupMenu(QListWidgetItem *item, bool preview)
 		action->setCheckable(true);
 		action->setChecked(ui->preview->Locked());
 
+		action = popup.addAction(
+			QTStr("Basic.MainMenu.Edit.ToggleScaling"),
```
Comment (RytoEX) [inline UI/window-basic-main.cpp]: Tab in twice for statement continuations (lines 3105-3106).

**Comment by RytoEX** [comment]
Style nitpick:  statement continuations are tabbed in twice.  Jim will probably make those edits to tabbing himself if you don't.

Out of curiosity.  What did you change on the first line?  When I compare them, they seem to be the same.

**Comment by Savestate2A03** [comment]
I am not sure, for some reason it says it's changed... Maybe something related to newlines? I just headed out to town and will be able to look further into it (and fix the indentation) when I get back. 
I appreciate the look-over!

**Comment by Savestate2A03** [comment]
Ok! Implemented the tab suggestions, couldn't figure out why the diff is bugging me about that first line in the locale.

**Comment by reboot** [comment]
The commit added a a UTF-8 byte order marker at be beginning, That is not visible on github. Whatever editor you used should be configured to save the file as one byte locale, like ASCII or ISO-8859-1. I would recommend always checking what you stage in git gui and if you see changes that you didn't do, just do not stage them and then revert the file after you commited.

File: UI/window-basic-main.cpp [hunk 3]
```diff
@@ -2573,10 +2578,17 @@ void OBSBasic::ResizePreview(uint32_t cx, uint32_t cy)
...
-			targetSize.width()  - PREVIEW_EDGE_SIZE * 2,
-			targetSize.height() - PREVIEW_EDGE_SIZE * 2,
-			previewX, previewY, previewScale);
+
+	if (ui->preview->Scaled()) {
+		GetScaleAndCenterPos(int(cx), int(cy),
+				targetSize.width() - PREVIEW_EDGE_SIZE * 2,
+				targetSize.height() - PREVIEW_EDGE_SIZE * 2,
+				previewX, previewY, previewScale);
+	} else {
```
Comment (RytoEX) [inline UI/window-basic-main.cpp]: Style nitpick:  There's a trailing space on line 2587.  Remove it.

**Comment by Savestate2A03** [comment]
Ah! Thanks for the info. I was using Sublime Text 3, and I'll look for the option. I'll see if I can fix it up and re-commit.

**Comment by RytoEX** [comment]
@reboot Nice catch!  I had just cloned the forked repo to check that and came to same conclusion.

**Comment by Savestate2A03** [comment]
Got the space, and removed the byte-order marker!

**Comment by reboot** [comment]
Shouldn't the preview be at least somehow scrollable, if it is not scaled? If my base resolution is the same as the monitor resolution then the preview will never fit into the window with all the UI around it. So in the current state it is only useful if your base resolution is lower then the monitor resolution, which is bad in my opinion because using the monitor resolution is the default for the base resolution.

**Comment by Savestate2A03** [comment]
I am used to streaming at 1280x720 on a larger monitor so the thought never crossed my mind.
I don't think there were ever scrollbars in OBS Classic, but that shouldn't be a reason to not add them in. I'll see what I can do!

**Comment by Savestate2A03** [comment]
Being my 1st time using QT, I don't know the ropes all too well. I had a prelim version working earlier today but it was (as far as I could tell) horribly hacked together. Any insight to how I might go about adding scrollbars the _right_ way?

**Comment by Savestate2A03** [comment]
I decided to not use scrollbars, as I could not figure out how to implement them properly. Instead, I went for the approach that I know After Effects uses, where if you hold down the space bar and click and drag, you can move the current view around. However, I've noticed that if the window is resized to be too small, the display is instead stretched (ignoring aspect ratio) to the full view area. I suspect this is something related to rendering something out of bounds? I am currently not sure how to figure that out... If someone could point me in the right direction, that'd be great. Otherwise, I'll keep poking around.

**Comment by Savestate2A03** [comment]
Figured it out! it was some integer overflow stuff. Changed a few things in the offset calculation to floats and it works perfectly.

File: UI/window-basic-preview.cpp [hunk 4]
```diff
@@ -377,6 +378,40 @@ void OBSBasicPreview::GetStretchHandleData(const vec2 &pos)
...
 
+void OBSBasicPreview::keyPressEvent(QKeyEvent *event)
+{
+	if (locked || scaled || event->isAutoRepeat()) {
+		OBSQTDisplay::keyPressEvent(event);
+		return;
+	}
+
+	switch (event->key()) {
+		case Qt::Key_Space:
```
Comment (RytoEX) [inline UI/window-basic-preview.cpp]: Style:  Don't indent case statements in a switch-case.  Align switch and its subordinate case labels.  Indent contents of each case.

File: UI/window-basic-preview.cpp [hunk 5]
```diff
@@ -377,6 +378,40 @@ void OBSBasicPreview::GetStretchHandleData(const vec2 &pos)
...
+
+void OBSBasicPreview::keyReleaseEvent(QKeyEvent *event)
+{
+	if (event->isAutoRepeat()) {
+		OBSQTDisplay::keyReleaseEvent(event);
+		return;
+	}
+
+	switch (event->key()) {
+		case Qt::Key_Space:
```
Comment (RytoEX) [inline UI/window-basic-preview.cpp]: Style:  Don't indent case statements in a switch-case.  Align switch and its subordinate case labels.  Indent contents of each case.

**Comment by Savestate2A03** [comment]
Got the switches indented properly, lemmie know if there's other stuff as well!

**Comment by style1925** [comment]
This is displayed in the Base (Canvas) Resolution?
Is it possible to Output (Scaling) Resolution?
It's different from OBS Classic.
Sorry about my poor English.

**Comment by Savestate2A03** [comment]
Hmm, I don't really see a reason for a fixed scale besides 1:1. What's the reason?

**Comment by style1925** [comment]
oops.
Maybe I might have misunderstood the function.
Whether it had been interpreted in this way, different Did you mean?
Left OBS Classic / right OBS Studio (You Pull requests)
Scale to fit mode = Scale Preview to Window (Enabled)
1:1 mode = Scale Preview to Window (Disabled)
If different, I'm sorry.

**Comment by Savestate2A03** [comment]
That is correct!

**Comment by Savestate2A03** [comment]
Well, while I don't see a particularly useful reason for the scaled resolution, I'll try and add it in! I guess it is kinda nice to be able to see the resolution you're streaming at.

**Comment by style1925** [comment]
Sorry for the late reply.
English is difficult…
Thank you for confirmation!

reason.
My output (Scaled) resolution is very small. (The bitrate of the site is limited. up to 480kb/s…)
So small, and the first to the scene editing. "scale to fit mode" 
Then to use here. "1:1 mode" 

This request, the ”Base (Canvas) Resolution” is used.
In OBS Classic, it is displayed in the Resolution Downscale. (1:1 mode)
It was very convenient!

**Comment by Savestate2A03** [comment]
Ah, I wasn't aware! My base resolution was always my output; I didn't know it was different in OBS Classic. Thanks for the heads up!

**Comment by Savestate2A03** [comment]
Here's what I'll try to have it look like when I'm finished with it.
image

**Comment by style1925** [comment]
You're welcome!

It is choice? (scale to window, canvas , output)
I think very good!

**Comment by Savestate2A03** [comment]
Yes (: they will be set based on your current canvas/output resolution. Just finished this actually, but I am without internet right now. I'm copying it to a USB drive and bringing it to my university later today to push the changes.

**Comment by Savestate2A03** [comment]
It seems that my USB drive fell out of my pocket while I was riding my bike to school. Looks like this will have to wait until tomorrow, sorry!

**Comment by style1925** [comment]
Good morning.

oh.
That’s sad...
tomorrow ok!

**Comment by Savestate2A03** [comment]
Ok! The update allows choosing between 3 scaling modes: Window, Canvas, and Output. 

As always, if I didn't get something, point it out and I'll make changes

**Comment by style1925** [comment]
It is amazing!
Thank you for changing.
It works properly!


Only one thing to worry about.
"Canvas / Output" Space position adjustment

procedure
Space position adjustment by either "Canvas / output"
Return to "Scale to Window"
Make it "Canvas / output" again. (The position has not been reset.)

the reason
"Canvas / output" has different resolution "Preview" is outside the screen.
In order to avoid it, I think that it is better to reset the position when "Canvas / Output" is selected.

**Comment by Savestate2A03** [comment]
An oversight on my part, thanks for pointing it out! I'll see if I can get that fixed by sometime tonight.

**Comment by Savestate2A03** [comment]
Looks like this will have to wait until tomorrow again. I installed visual studio on a school computer to fix it up, but it required a restart after installation and these computers have deep freeze, so that's not happening haha.

**Comment by Savestate2A03** [comment]
Looks like that's everything; lemmie know if there's anything else I should do, whether it be code or formatting... Otherwise, I think it's ready to pull!

**Comment by jp9000** [comment]
This is one of the best pull requests I've ever received.  Everybody helped work together and shared their opinions, and it made for a great pull request for a great feature.  Well done.

Manually merging with three minor changes:
1.) There was some trailing whitespace in some of your code.  You can use git's diff feature to view that trailing whitespace.  Git will show you that whitespace as big red blocks.
2.) The code styling was fine, though I only changed one thing, the part where you had the expression of an "if" statement broken in to multiple lines.  Personally I don't like to begin lines with ||, I'd rather have them on the end.  Just changed that to make the code more consistent.
3.) There were no notes in the commit itself.  It was only a commit title.  Your commit title was a bit long as well, and probably indicative that you could have shortened it and expanded upon it in the detailed part of the commit message.

Other than that it's a fantastic feature, and also is a parity feature with OBS classic that we've been needing.  Thank you very much!

**Comment by Savestate2A03** [comment]
I've used your software so much, it's a small way of saying thank you! I also appreciate the tips; I'll keep them in mind when working on OBS in the future.

**Comment by reboot** [comment]
It works really well. The only thing i'm wondering about is if it is really a good idea that you can not scroll the preview when it is locked. Locking could mean that you can not change the preview at all, so that would include that you can not change the current visible area. But the original intent of locking might only have been that you can not accidently edit the scene. In that case I would say that it would still be useful if you can scroll the preview to look at different parts of it. Even if the preview is locked, you can still move sources using edit transformation, if you want to check if a position is correct you can not look at different parts of the preview for that.

**Comment by Savestate2A03** [comment]
You make a good point. It probably wouldn't be very hard to fix, but I don't know how to go about fixing it since it was already pulled. Would I make another pull request, and then Jim can rebase it or -

Also it would probably be another day before I could get to it since I'm still without internet at my apartment, but I'm sure it's simple enough for someone else to do it in the time.

**Comment by reboot** [comment]
No need to rush, it is not broken, it is just up for discussion if this should be improved. This pull request is in master, just start a new branch from the current master with the improvement and make a new PR.

**Comment by Savestate2A03** [comment]
Sweet, can do!

**Pull request closed without merging**